File: packages/typescript/examples/09_testing.ts
/**
 * Testing your agents.
 *
 * Run: npx ts-node examples/09_testing.ts
 */
import { Agent, AgentTestClient } from '../src';

const agent = new Agent({ name: 'Calculator', description: 'Math operations' });

agent.skill('add', async ({ a, b }: { a: number; b: number }) => a + b);
agent.skill('multiply', async ({ a, b }: { a: number; b: number }) => a * b);
agent.skill('divide', async ({ a, b }: { a: number; b: number }) => {
  if (b === 0) return { error: 'Cannot divide by zero' };
  return { result: a / b };
});

async function runTests() {
  console.log('Running tests...\n');
  const client = new AgentTestClient(agent);

  // Test add
  const sum = await client.call('add', { a: 2, b: 3 });
  console.assert(sum.data === 5, 'Expected 5');
  console.log('test_add passed');

  // Test multiply
  const product = await client.call('multiply', { a: 4, b: 5 });
  console.assert(product.data === 20, 'Expected 20');
  console.log('test_multiply passed');

  // Test divide
  const quotient = await client.call('divide', { a: 10, b: 2 });
  console.assert((quotient.data as any).result === 5, 'Expected 5');
  console.log('test_divide passed');

  // Test divide by zero
  const error = await client.call('divide', { a: 10, b: 0 });
  console.assert('error' in (error.data as any), 'Expected error');
  console.log('test_divide_by_zero passed');

  // Test list skills
  const skills = client.listSkills();
  console.assert(skills.includes('add'), 'Expected add skill');
  console.assert(skills.includes('multiply'), 'Expected multiply skill');
  console.assert(skills.includes('divide'), 'Expected divide skill');
  console.log('test_list_skills passed');

  // Test get agent card
  const card = client.getAgentCard();
  console.assert(card.name === 'Calculator', 'Expected Calculator');
  console.assert(card.skills?.length === 3, 'Expected 3 skills');
  console.log('test_get_agent_card passed');

  console.log('\nAll tests passed!');
}

runTests();

File: packages/python/examples/09_testing.py
"""
Example: Testing your agents.

A2A Lite includes a TestClient that makes testing trivial.

Run: python examples/09_testing.py
"""
from a2a_lite import Agent, AgentTestClient


# Create a simple agent
agent = Agent(name="Calculator", description="Math operations")


@agent.skill("add")
async def add(a: int, b: int) -> int:
    return a + b


@agent.skill("multiply")
async def multiply(a: int, b: int) -> int:
    return a * b


@agent.skill("divide")
async def divide(a: float, b: float) -> dict:
    if b == 0:
        return {"error": "Cannot divide by zero"}
    return {"result": a / b}


# Testing is this simple:
def test_add():
    client = AgentTestClient(agent)
    result = client.call("add", a=2, b=3)
    assert result == 5
    print("✅ test_add passed")


def test_multiply():
    client = AgentTestClient(agent)
    result = client.call("multiply", a=4, b=5)
    assert result == 20
    print("✅ test_multiply passed")


def test_divide():
    client = AgentTestClient(agent)
    result = client.call("divide", a=10, b=2)
    assert result.data["result"] == 5.0
    print("✅ test_divide passed")


def test_divide_by_zero():
    client = AgentTestClient(agent)
    result = client.call("divide", a=10, b=0)
    assert "error" in result.data
    print("✅ test_divide_by_zero passed")


def test_list_skills():
    client = AgentTestClient(agent)
    skills = client.list_skills()
    assert "add" in skills
    assert "multiply" in skills
    assert "divide" in skills
    print("✅ test_list_skills passed")


if __name__ == "__main__":
    # Run all tests
    print("Running tests...\n")

    test_add()
    test_multiply()
    test_divide()
    test_divide_by_zero()
    test_list_skills()

    print("\n🎉 All tests passed!")

    # Optionally run the server
    # agent.run(port=8787)
